### DISCUSSION TITLE
Looking for the howto use Micropython in docker

### DISCUSSION THREAD

**Discussion by Josverl** [micropython/micropython#15413] [Using MicroPython] [answered]
I am looking to use a dockerized micropython in GHA / pytest to verify that the micropython type-stubs and typings modules do not interfere with micropython at runtime, and docker seems to be a simple way to do that by make use of work already done.

a snippet from the test:

```python
    cmd = f"docker run -u 1000 -v .:/code --rm micropython/unix:{mp_version} micropython {check_file}"
    result = subprocess.run(cmd, shell=True, cwd=snip_path, text=True, capture_output=True)
    error = [line for line in result.stdout.split("\n") if "Traceback" not in line]
    assert result.returncode == 0, error
``` 
all good, 
Testing runs fine, and take about 500ms each after the initial pull, which is much better than any manual testing, while leaving room for improvement.

but its a bit of a black box , and I´d like to understand better how this is setup internally, and how to inject modules into the `/libs` folder.
looking at `sys.path` I see a few things I´d like to understand better 

``` bash
docker run -u 1000 -it --rm micropython/unix:latest 
MicroPython v1.23.0 on 2024-06-13; linux [GCC 12.2.0] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import sys;print(sys.path)
['', '.frozen', '//.micropython/lib', '/usr/lib/micropython']
```

Both folders cannot be located in the image, are they intended as placeholder that can optionally be mounted ?
running outside docker `sys.path` is `['', '.frozen', '/home/jos/.micropython/lib', '/usr/lib/micropython']` so it seems that I should pass in some locations or variables 

I tried looking at the docker image but  could not figure out whether i need to look at configuration for my host(s), docker, or micropython.

@mattytrentini , I think you may know ...

> as an aside: 
Searching for `"micropython/unix"` in Github does _not_ seem to do an exact text search :-(

**Comment by mattytrentini** [comment, marked as answer]
Good to see the containers getting some love! :) 

In case it's not obvious, to create these containers I just build the default unix build - and those are the standard paths for the unix build...

The `//.micropython/lib` is the only interesting one - if I run as root (inside WSL) it looks different:

```bash
> docker run -it --rm micropython/unix
MicroPython v1.23.0 on 2024-06-13; linux [GCC 12.2.0] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import sys; print(sys.path)
['', '.frozen', '/root/.micropython/lib', '/usr/lib/micropython']
>>>
```

At a guess, I'd say that `//.micropython/lib` is the home folder of the user that docker runs under - so perhaps your user doesn't have a home folder?

In any case, you can have multiple volume mounts and you can configure `sys.path` to whatever you need. 

Does that help?

**Reply by Josverl** [reply]
Thanks for the quick reply 

This run was on ubuntu. 

But it may be a side effect of the `-u 1000` that I borrowed from some of your other examples.
That would explain that $HOME is not set and may be an option to mount my lib folder under one of these paths.

**Comment by Josverl** [comment]
Indeed defining $HOME correctly sets the 3rd path :
```
docker run -u 1000 --rm -e HOME=/foo  micropython/unix:latest micropython -c "import sys;print(sys.path)"
['', '.frozen', '/foo/.micropython/lib', '/usr/lib/micropython']
```
and then I can mount my lib folder in that location, without a need to change the sys.path

```
docker run -u 1000 --rm -e HOME=/foo -v .:/code -v ./lib:/foo/.micropython/lib micropython/unix:latest micropython check_basics.py
```

**Reply by mattytrentini** [reply]
Glad to hear it's sorted out! Please let me know if you have any ideas about improvements to the containers - I'm publishing them manually at the moment but at some point I'll raise a PR to build/publish with each commit/tag.
